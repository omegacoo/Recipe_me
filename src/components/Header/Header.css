/* Mobile Phone Portrait */
.Header {
    height: 7vh;
    background-image: url('../../assets/marble.jpg');
    background-size: cover;
    display: grid;
    grid-template-columns: 7vh;
    grid-template-rows: 10vw 80vw 10vw;
}

.logo {
    margin: 1vw;
    height: 11vh;
    z-index: 2;
}

.login_icon {
    grid-column: 13;
    width: 5vh;
    font-size: 5vh;
    color: black;
    margin-top: 1vh;
}

.popupContainer {
    position: relative;
    display: inline-block;
}

.Header .popup {
    z-index: 2;
    background-image: url('../../assets/postit.png');
    background-size: 60vw 100%;
    background-repeat: no-repeat;
    color: black;
    padding: 2vh;
    position: absolute;
    right: -87vw;
    top: -5vh;
    width: 60vw;
    font-weight: bold;
}

.popup input {
    width: 50vw;
    border: none;
    border-bottom: 1px solid black;
    padding-left: 5px;
    padding-right: 5px;
    background-color: rgb(254, 254, 72);
}

.popup ul {
    margin-top: 3vh;
    text-align: center;
}

.popup li {
    list-style: none;
    margin: 4vh;
    margin-bottom: 0;
}

.navLink {
    text-decoration: none;
    border: 2px solid black;
    border-radius: 15px;
    padding: 1vh;
    background-color: white;
    font-weight: bold;
}

.closePopup {
    background-color: red;
    color: #F0EAD6;
    border-radius: 50%;
    font-size: 3vh;
    position: absolute;
    right: 6vw;
    top: 1vh;
    box-shadow: 3px 3px 5px black;
}

.forgotInfo {
    font-size: 1.7vh;
}

.login_button {
    display: block;
    width: 50%;
    margin-left: auto;
    margin-right: auto;
    margin-top: 1vh;
    margin-bottom: 1vh;
    background-color: #F0EAD6;
    border: 1px solid black;
    border-radius: 15px;
    padding: 1vh;
    font-weight: bolder;
    box-shadow: 5px 5px 5px;
}

.hamburger {
    grid-column: 13;
    width: 30px;
    font-size: 4vh;
    color: black;
    margin-top: 1.5vh;
}

.log_out {
    font-size: 3vh;
    position: absolute;
    right: 2vw;
    top: 1vh;
    text-shadow: 0 -1px 0 #000;
}

/* iPhone X 375x812 Portrait */
@media only screen and (min-width: 375px) and (min-height: 812px) {
    .Header .popup {
        top: -4vh;
        right: -85vw;
    }
}

/* iPad 768x1024 Portrait */
@media only screen and (min-width: 768px) and (min-height: 1024px) {
    .login_icon {
        width: 2vh;
    }

    .Header .popup {
        top: -7vh;
        right: -90vw;
        width: 40vw;
        height: 60vh;
        background-size: 40vw;
    }

    .Header .popup label {
        font-size: 3vw;
    }

    .Header .popup input {
        width: 30vw;
        height: 3vh;
        font-size: 2vh;
        margin-bottom: 2vh;
    }

    .login_button {
        font-size: 3vw;
        margin-top: 0;
    }

    .navLink {
        font-size: 3vh;
    }

    .closePopup {
        right: 4vw;
    }
}