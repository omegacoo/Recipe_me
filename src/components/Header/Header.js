import React from 'react';
import { Link } from 'react-router-dom';
import config from '../../config';
import StoreContext from '../../StoreContext';

import logo from '../../assets/pantry.png';

import './Header.css';

export default class Header extends React.Component {
    static contextType = StoreContext;

    state = {
        username: '',
        password: '',
        userId: null,
        logingIn: false,
        error: null,
    };

    toggleLogin = () => {
        this.setState({
            logingIn: !this.state.logingIn
        });
    };

    handleLoginSubmit = e => {
        e.preventDefault();

        const myHeaders = new Headers();
        myHeaders.append("Content-Type", "application/json");
        const myBody = JSON.stringify({ "user_name": this.state.username, "password": this.state.password });

        fetch(config.API_ENDPOINT + '/api/auth/login',  { method: 'POST', body: myBody, headers: myHeaders })
            .then(res => {
                if(res.status === 401){
                    throw new Error('Incorrect user_name or password') 
                };
                this.setState({
                    error: null
                });
                
                const Xtoken = res.headers.get('X-token');
                const userId = res.headers.get('user_id');

                this.setState({
                    userId
                });
                
                document.cookie = `token=${Xtoken}`;
                                 
                return res.text()
            })
            .then(resText => {    
                this.context.onLogin(this.state.username, this.state.userId);
                this.toggleLogin();
                this.props.history.push('/pantry');
            })
            .catch(err => {
                this.setState({
                    error: err.message
                });
            })
            
    };

    logOut = () => {
        this.setState({
            user_name: '',
            password: '',
            userId: null,
            logingIn: false
        },
            this.context.onLogout()
        );
    };

    handleUsernameChange = e => {
        this.setState({
            username: e.target.value,
            error: null
        });
    };

    handlePasswordChange = e => {
        this.setState({
            password: e.target.value,
            error: null
        });
    };

    renderPopup = () => {
        if(!this.context.loggedIn && this.state.logingIn){
            return(
                <div className='popupContainer'>
                    <span className='popup'>
                        <i className='closePopup far fa-times-circle' onClick={this.toggleLogin} />
                        <form 
                            className='login_form'
                            onSubmit={this.handleLoginSubmit}
                        >
                            <label htmlFor='login_username'>Username: </label>
                            <input 
                                value={this.state.username}
                                onChange={this.handleUsernameChange}
                                type='text'
                                autoFocus
                            />
                            <label htmlFor='login_password'>Password: </label>
                            <input
                                value={this.state.password}
                                onChange={this.handlePasswordChange}
                                type='password'
                            />
                            {this.state.error ? <h3 id='Login_error'>{this.state.error}</h3> : null}
                            <button className='login_button' type='submit'>login</button>
                        </form>
                    </span>
                </div>
            )
        };
    };

    renderLoginOrLogout = () => {
        if(!this.context.loggedIn && !this.state.logingIn){
            return <i className='fas fa-key login_icon' onClick={this.toggleLogin} />
        } else if(this.context.loggedIn && !this.state.logingIn){
            return <Link className='log_out' onClick={this.logOut} to={'/'}>logout</Link>
        };    
    };

    render(){
        return(
        <header className='Header'>
            <img 
                src={logo} 
                className='logo'
                alt='Pantry Buddy logo'
                onClick={() => this.props.history.push('/')}  
            />
            { this.renderLoginOrLogout() }
            { this.renderPopup() }
        </header>
        );
    };
};