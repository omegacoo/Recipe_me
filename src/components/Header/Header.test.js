import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import Header from './Header';

describe('Header component', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');

        ReactDOM.render(
            <BrowserRouter>
                <Header location={ { pathname:'/' } }/>
            </BrowserRouter>, 
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders UI correctly', () => {
        expect(
        renderer.create(
            <BrowserRouter>
                <Header  location={ { pathname:'/' } }/>
            </BrowserRouter>).toJSON()
        ).toMatchSnapshot();
    });
});