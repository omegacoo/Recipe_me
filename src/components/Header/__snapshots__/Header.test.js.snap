// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders UI correctly 1`] = `
<header
  className="Header"
>
  <img
    alt="Pantry Buddy logo"
    className="logo"
    onClick={[Function]}
    src="pantry.png"
  />
  <a
    className="log_out"
    href="/"
    onClick={[Function]}
  >
    logout
  </a>
</header>
`;
